### PR TITLE
[SPARK-25492][TEST] Refactor WideSchemaBenchmark to use main method

### PR DESCRIPTION
**Pull request by wangyum**
## What changes were proposed in this pull request?

Refactor `WideSchemaBenchmark` to use main method.
1. use `spark-submit`:
```console
bin/spark-submit --class  org.apache.spark.sql.execution.benchmark.WideSchemaBenchmark --jars ./core/target/spark-core_2.11-3.0.0-SNAPSHOT-tests.jar ./sql/core/target/spark-sql_2.11-3.0.0-SNAPSHOT-tests.jar
```

2. Generate benchmark result:
```console
SPARK_GENERATE_BENCHMARK_FILES=1 build/sbt "sql/test:runMain org.apache.spark.sql.execution.benchmark.WideSchemaBenchmark"
```

## How was this patch tested?

manual tests

### CODE REVIEW

File: sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/WideSchemaBenchmark.scala [hunk 1]
```diff
@@ -17,22 +17,19 @@
...
  * Benchmark for performance with very wide and nested DataFrames.
- * To run this:
- *  build/sbt "sql/test-only *WideSchemaBenchmark"
- *
- * Results will be written to "sql/core/benchmarks/WideSchemaBenchmark-results.txt".
+ * To run this benchmark:
+ * 1. without sbt: bin/spark-submit --class <this class> <spark sql test jar>
+ * 2. build/sbt "sql/test:runMain <this class>"
+ * 3. generate result: SPARK_GENERATE_BENCHMARK_FILES=1 build/sbt "sql/test:runMain <this class>"
+ *    Results will be written to "benchmarks/WideSchemaBenchmark-results.txt".
```
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/WideSchemaBenchmark.scala]: Could you fix doc generation failure?

Comment (wangyum) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/WideSchemaBenchmark.scala]: Thanks @dongjoon-hyun. Actually I'm waiting for https://github.com/apache/spark/pull/22484. I want to move  `withTempDir()` to  `RunBenchmarkWithCodegen.scala`.

File: core/src/test/scala/org/apache/spark/benchmark/BenchmarkBase.scala [hunk 2]
```diff
@@ -48,15 +48,11 @@ abstract class BenchmarkBase {
       if (!file.exists()) {
         file.createNewFile()
       }
-      output = Some(new FileOutputStream(file))
+      output = Option(new FileOutputStream(file))
```
Comment (dongjoon-hyun) [inline core/src/test/scala/org/apache/spark/benchmark/BenchmarkBase.scala]: This looks like irrelevant pig-back.

Comment (wangyum) [reply core/src/test/scala/org/apache/spark/benchmark/BenchmarkBase.scala]: Change here because: https://github.com/apache/spark/pull/22443#discussion_r221181428

Comment (dongjoon-hyun) [reply core/src/test/scala/org/apache/spark/benchmark/BenchmarkBase.scala]: IIUC, @HyukjinKwon meant `when you need to touch this file`.

Comment (wangyum) [reply core/src/test/scala/org/apache/spark/benchmark/BenchmarkBase.scala]: I am worried that I will forget it after a long time, so I am changing this time. I should revert it?

Comment (dongjoon-hyun) [reply core/src/test/scala/org/apache/spark/benchmark/BenchmarkBase.scala]: Why do you replace `Some` to `Option`? Are you worrying `new FileOutputStream(file)` becomes `null`?

Comment (HyukjinKwon) [reply core/src/test/scala/org/apache/spark/benchmark/BenchmarkBase.scala]: My point was that there's no point of checking `null` below from my cursory look. If there's no chance that it becomes `null`, we can leave it `Some` and remove `null` check below.

File: sql/core/benchmarks/WideSchemaBenchmark-results.txt [hunk 3]
```diff
@@ -1,117 +1,145 @@
...
-2500 cols x 40 rows (exec in-mem)              326 /  330          0.3        3260.6       0.0X
-2500 cols x 40 rows (read parquet)             831 /  839          0.1        8305.8       0.0X
-2500 cols x 40 rows (write parquet)            237 /  245          0.4        2372.6       0.1X
-
-Java HotSpot(TM) 64-Bit Server VM 1.8.0_92-b14 on Mac OS X 10.11.6
-Intel(R) Core(TM) i7-4980HQ CPU @ 2.80GHz
+1 cols x 100000 rows (read in-mem)              22 /   25          4.6         219.4       1.0X
+1 cols x 100000 rows (exec in-mem)              22 /   28          4.5         223.8       1.0X
+1 cols x 100000 rows (read parquet)             45 /   49          2.2         449.6       0.5X
+1 cols x 100000 rows (write parquet)           204 /  223          0.5        2044.4       0.1X
```
Comment (dongjoon-hyun) [inline sql/core/benchmarks/WideSchemaBenchmark-results.txt]: This might be a little regression on Parquet writer from Spark 2.1.0 (SPARK-17335).

cc @cloud-fan and @gatorsmile , @rdblue

File: sql/core/benchmarks/WideSchemaBenchmark-results.txt [hunk 4]
```diff
@@ -1,117 +1,145 @@
...
+100 cols x 1000 rows (read in-mem)              26 /   28          3.9         255.8       0.9X
+100 cols x 1000 rows (exec in-mem)              32 /   35          3.1         319.3       0.7X
+100 cols x 1000 rows (read parquet)             45 /   52          2.2         445.9       0.5X
+100 cols x 1000 rows (write parquet)           275 /  536          0.4        2746.1       0.1X
+2500 cols x 40 rows (read in-mem)              261 /  434          0.4        2607.3       0.1X
+2500 cols x 40 rows (exec in-mem)              624 /  701          0.2        6240.5       0.0X
+2500 cols x 40 rows (read parquet)             196 /  301          0.5        1963.4       0.1X
+2500 cols x 40 rows (write parquet)            687 / 1049          0.1        6870.6       0.0X
```
Comment (dongjoon-hyun) [inline sql/core/benchmarks/WideSchemaBenchmark-results.txt]: The difference between `best` and `average` is too high in line 32 and line 33.
I'll try to run this on EC2, too.

File: sql/core/benchmarks/WideSchemaBenchmark-results.txt [hunk 5]
```diff
@@ -1,117 +1,145 @@
...
-
-Java HotSpot(TM) 64-Bit Server VM 1.8.0_92-b14 on Mac OS X 10.11.6
-Intel(R) Core(TM) i7-4980HQ CPU @ 2.80GHz
+1 deep x 100000 rows (read in-mem)              20 /   24          5.1         197.9       1.0X
+1 deep x 100000 rows (exec in-mem)              23 /   28          4.4         227.8       0.9X
+1 deep x 100000 rows (read parquet)             50 /   58          2.0         500.1       0.4X
+1 deep x 100000 rows (write parquet)           195 /  219          0.5        1945.1       0.1X
+100 deep x 1000 rows (read in-mem)              39 /   57          2.5         393.1       0.5X
+100 deep x 1000 rows (exec in-mem)             480 /  556          0.2        4795.7       0.0X
+100 deep x 1000 rows (read parquet)           7943 / 7950          0.0       79427.5       0.0X
```
Comment (dongjoon-hyun) [inline sql/core/benchmarks/WideSchemaBenchmark-results.txt]: Ur, @wangyum . Is this 4 times slower than before?

cc @dbtsai .

File: sql/core/benchmarks/WideSchemaBenchmark-results.txt (hunk 3 above)
Comment (cloud-fan) [reply sql/core/benchmarks/WideSchemaBenchmark-results.txt]: I have no idea how this happens. Can you create a JIRA ticket to investigate this regression?

Comment (wangyum) [reply sql/core/benchmarks/WideSchemaBenchmark-results.txt]: May be a parquet issue. I found that the binary write performance is a little worse after upgrading to parquet 1.10.0: https://github.com/apache/parquet-mr/pull/505. I will verify it later.

Comment (dongjoon-hyun) [reply sql/core/benchmarks/WideSchemaBenchmark-results.txt]: The following [EC2 result](https://github.com/wangyum/spark/pull/19) shows the consistent ratio like Spark 2.1.0. The result on Mac seemed to be unstable for some unknown reason like https://github.com/apache/spark/pull/22501#discussion_r226440992. 
```scala
1 cols x 100000 rows (read parquet)             61 /   70          1.6         610.2       0.6X
1 cols x 100000 rows (write parquet)           209 /  233          0.5        2086.1       0.2X
```

Comment (rdblue) [reply sql/core/benchmarks/WideSchemaBenchmark-results.txt]: @dongjoon-hyun, so you are saying that it doesn't appear that there is a performance regression, right?

Comment (dongjoon-hyun) [reply sql/core/benchmarks/WideSchemaBenchmark-results.txt]: For this part, right, @rdblue . I guess so.
After merging EC2 result to @wangyum 's PR, I'll compare the numbers one by one once again.

**Comment by cloud-fan** [comment]
thank you guys for refreshing the benchmarks and results! It's very helpful.

If possible, can we post the perf regressions we found in the umbrella JIRA? Then people can see if the perf regression is reasonable(if we have addressed it) or investigate how the regression was introduced.

Thanks!

**Comment by cloud-fan** [comment]
seems jenkins is broken, cc @shaneknapp 
```
Command "/tmp/tmp.JfFHaoRFPU/3.5/bin/python -c "import setuptools, tokenize;__file__='/home/jenkins/workspace/SparkPullRequestBuilder/python/setup.py';f=getattr(tokenize, 'open', open)(__file__);code=f.read().replace('\r\n', '\n');f.close();exec(compile(code, __file__, 'exec'))" develop --no-deps" failed with error code 1 in /home/jenkins/workspace/SparkPullRequestBuilder/python/
You are using pip version 10.0.1, however version 18.1 is available.
You should consider upgrading via the 'pip install --upgrade pip' command.
Cleaning up temporary directory - /tmp/tmp.JfFHaoRFPU
[error] running /home/jenkins/workspace/SparkPullRequestBuilder/dev/run-pip-tests ; received return code 1
```

**Comment by dongjoon-hyun** [comment]
@cloud-fan After updating on EC2, almost ratio and values looks more stable and reasonable for now. The following two are noticeable changes, but it looks like Parquet Writer improvement (instead of regression).

**1. Read/Write ratio is reverted (`0.8` -> `1.7`)**
I'm not sure but Parquet writer for `deep
```scala
- 128 x 8 deep x 1000 rows (read parquet)         69 /   74          1.4         693.9       0.2X
- 128 x 8 deep x 1000 rows (write parquet)        78 /   83          1.3         777.7       0.2X
+ 128 x 8 deep x 1000 rows (read parquet)        351 /  379          0.3        3510.3       0.1X
+ 128 x 8 deep x 1000 rows (write parquet)       199 /  203          0.5        1988.3       0.2X
```

**2. Read/Write ratio is changed noticeably (`4.6` -> `8.3`)**
```scala
- 1024 x 11 deep x 100 rows (read parquet)        426 /  433          0.2        4263.7       0.0X
- 1024 x 11 deep x 100 rows (write parquet)        91 /   98          1.1         913.5       0.1X
+ 1024 x 11 deep x 100 rows (read parquet)       2063 / 2078          0.0       20629.2       0.0X
+ 1024 x 11 deep x 100 rows (write parquet)       248 /  266          0.4        2475.1       0.1X
```

Since this is the first attempt to track this and the previous result is too old, there exists some obvious limitation during comparison. From Spark 2.4.0, we can get a consistent compasison instead of `different` personal mac.

**Review by dongjoon-hyun — APPROVED** [review]

**Comment by dongjoon-hyun** [comment]
Retest this please.

**Comment by cloud-fan** [comment]
retest this please

**Comment by shaneknapp** [comment]
@cloud-fan  --  pip isn't broken...  the actual error is found right above what you cut and pasted:

`    UnicodeDecodeError: 'ascii' codec can't decode byte 0xc2 in position 2719: ordinal not in range(128)`

i won't be able to look any deeper in to this until at least tomorrow at the earliest.

**Comment by HyukjinKwon** [comment]
I guess it's related with pip packaging tho. 

```
    Traceback (most recent call last):
      File "<string>", line 1, in <module>
      File "/home/jenkins/workspace/SparkPullRequestBuilder/python/setup.py", line 224, in <module>
        'Programming Language :: Python :: Implementation :: PyPy']
      File "/tmp/tmp.R2Y98bevgD/3.5/lib/python3.5/site-packages/setuptools/__init__.py", line 140, in setup
        return distutils.core.setup(**attrs)
      File "/tmp/tmp.R2Y98bevgD/3.5/lib/python3.5/distutils/core.py", line 148, in setup
        dist.run_commands()
      File "/tmp/tmp.R2Y98bevgD/3.5/lib/python3.5/distutils/dist.py", line 955, in run_commands
        self.run_command(cmd)
      File "/tmp/tmp.R2Y98bevgD/3.5/lib/python3.5/distutils/dist.py", line 974, in run_command
        cmd_obj.run()
      File "/tmp/tmp.R2Y98bevgD/3.5/lib/python3.5/site-packages/setuptools/command/develop.py", line 38, in run
        self.install_for_development()
      File "/tmp/tmp.R2Y98bevgD/3.5/lib/python3.5/site-packages/setuptools/command/develop.py", line 154, in install_for_development
        self.process_distribution(None, self.dist, not self.no_deps)
      File "/tmp/tmp.R2Y98bevgD/3.5/lib/python3.5/site-packages/setuptools/command/easy_install.py", line 729, in process_distribution
        self.install_egg_scripts(dist)
      File "/tmp/tmp.R2Y98bevgD/3.5/lib/python3.5/site-packages/setuptools/command/develop.py", line 189, in install_egg_scripts
        script_text = strm.read()
      File "/tmp/tmp.R2Y98bevgD/3.5/lib/python3.5/encodings/ascii.py", line 26, in decode
        return codecs.ascii_decode(input, self.errors)[0]
    UnicodeDecodeError: 'ascii' codec can't decode byte 0xc2 in position 2719: ordinal not in range(128)
    
```

It's from setup.py

**Comment by kiszk** [comment]
I am looking at each commit from the latest to old at https://github.com/apache/spark/commits/master

**Comment by HyukjinKwon** [comment]
Thanks. It might rather more be related to external factors.

**Comment by kiszk** [comment]
Is [this](https://github.com/apache/spark/pull/22748#issuecomment-431512558) the oldest test failure related to this type of failure?

**Comment by HyukjinKwon** [comment]
Yup, I made a fix https://github.com/apache/spark/pull/22782

**Comment by kiszk** [comment]
Thanks, I found `0xc2` in `docker-image-tool.sh`. I will put my finding into #22782

**Comment by dongjoon-hyun** [comment]
Retest this please.

**Comment by dongjoon-hyun** [comment]
Merged to master.

**Comment by dongjoon-hyun** [comment]
Thank you, @wangyum and all!

**Comment by MaxGekk** [comment]
Any ideas how much memory does this benchmark require. I set 32Gb but it crashes with OOM on the master:
```
export SBT_OPTS="-Xmx32g -XX:MaxPermSize=512M -XX:ReservedCodeCacheSize=512m"
```
```
SPARK_GENERATE_BENCHMARK_FILES=1 build/sbt "sql/test:runMain org.apache.spark.sql.execution.benchmark.WideSchemaBenchmark"
```
```
[error] Caused by: java.lang.reflect.InvocationTargetException
[error] 	at sun.reflect.GeneratedConstructorAccessor8.newInstance(Unknown Source)
[error] 	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
[error] 	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
[error] 	at org.apache.spark.sql.catalyst.trees.TreeNode.$anonfun$makeCopy$7(TreeNode.scala:468)
[error] 	at org.apache.spark.sql.catalyst.trees.CurrentOrigin$.withOrigin(TreeNode.scala:72)
[error] 	at org.apache.spark.sql.catalyst.trees.TreeNode.$anonfun$makeCopy$1(TreeNode.scala:467)
[error] 	at org.apache.spark.sql.catalyst.errors.package$.attachTree(package.scala:52)
[error] 	... 132 more
[error] Caused by: java.lang.OutOfMemoryError: GC overhead limit exceeded
[error] 	at java.util.Arrays.copyOf(Arrays.java:3332)
[error] 	at java.lang.AbstractStringBuilder.ensureCapacityInternal(AbstractStringBuilder.java:124)
[error] 	at java.lang.AbstractStringBuilder.append(AbstractStringBuilder.java:448)
[error] 	at java.lang.StringBuilder.append(StringBuilder.java:136)
[error] 	at org.apache.spark.sql.types.StructType.catalogString(StructType.scala:411)
[error] 	at org.apache.spark.sql.types.StructType.$anonfun$catalogString$1(StructType.scala:410)
[error] 	at org.apache.spark.sql.types.StructType$$Lambda$2439/60886146.apply(Unknown Source)
[error] 	at scala.collection.TraversableLike.$anonfun$map$1(TraversableLike.scala:238)
```

**Comment by dongjoon-hyun** [comment]
It sounds like a regression, @MaxGekk . This test suite was executed with the example in the code without any additional information like `export SBT_OPTS="-Xmx32g -XX:MaxPermSize=512M -XX:ReservedCodeCacheSize=512m"`.

**Comment by dongjoon-hyun** [comment]
FYI, this was tested in this PR, `Oct 19, 2018`, initially. The latest run was 854a0f752e , `Oct 3, 2019` in the `[SPARK-29320][TESTS] Compare `sql/core` module in JDK8/11 (Part 1)` PR.

**Comment by MaxGekk** [comment]
I have rechecked the benchmark on the recent master - it is still failing with OOM. I opened the JIRA ticket: https://issues.apache.org/jira/browse/SPARK-30429
